### PR TITLE
Feature/batch reward

### PR DESCRIPTION
**Pull request by alexandresablayrolles**

### CODE REVIEW

File: python/spinn/fat_stack.py [hunk 1]
```diff
@@ -360,8 +361,10 @@ def run(self, print_transitions=False, run_internal_parser=False,
 
             transition_acc = F.accuracy(
                 hyp_acc, truth_acc.astype(np.int32))
-            transition_loss = F.softmax_cross_entropy(
-                hyp_xent, truth_xent.astype(np.int32),
+            rewards = np.ndarray((truth_xent.shape[0]))
+            rewards[:] = 1.0 / truth_xent.shape[0]
+            transition_loss = batch_weighted_softmax_cross_entropy(
+                hyp_xent, truth_xent.astype(np.int32), rewards,
```
Comment (mrdrozdov) [inline python/spinn/fat_stack.py]: @alexandresablayrolles I realized that hyp_xent are the probabilities aka the output of F.softmax(transitions_hyp). Since batch_weighted_softmax_cross_entropy does a softmax also (I think), then shouldn't we be using the transitions_hyp here instead?

File: python/spinn/util/batch_softmax_cross_entropy.py [hunk 2]
```diff
@@ -111,59 +111,33 @@ def backward_cpu(self, inputs, grad_outputs):
...
-            'T y, S t, raw T coeff, S n_channel, S n_unit',
-            'T gx',
-            '''
-               const int c = (i / n_unit % n_channel);
-               gx = (t == -1) ? 0 : (coeff[0] * (y - (c == t)));
-            ''',
-            'softmax_crossent_bwd')(
-                y, cupy.expand_dims(t, 1), coeff, x.shape[1], n_unit)
-        return gx, None
+        gx *= numpy.tile(gloss*self._coeff, (gx.shape[1], 1)).T
```
Comment (mrdrozdov) [inline python/spinn/util/batch_softmax_cross_entropy.py:114]: why is this not just `gloss * self._coeff`?

File: python/spinn/fat_stack.py [hunk 3]
```diff
@@ -373,6 +385,34 @@ def run(self, print_transitions=False, run_internal_parser=False,
...
+
+        statistics = [
+            F.squeeze(F.concat([F.expand_dims(ss, 1) for ss in s], axis=0))
+            if isinstance(s[0], Variable) else
+            np.array(reduce(lambda x, y: x + y.tolist(), s, []))
+            for s in statistics]
+
+        hyp_acc, truth_acc, hyp_xent, truth_xent = statistics
+
+        rewards = np.repeat(rewards, (hyp_xent.shape[0] / rewards.shape[0]))
```
Comment (mrdrozdov) [inline python/spinn/fat_stack.py]: @alexandresablayrolles various solutions to get the sort of granularity you need w. hyp_xent. Right now it's like this:

```
if not use_skips:
  hyp_xent = hyp_xent[cant_skip]
mem['hyp_xent'] = hyp_xent
mems.append(mem)
```

Comment (mrdrozdov) [reply python/spinn/fat_stack.py]: Simplest thing would be array of arrays for memories. This could look like:

```
# init mems
mems = {}
mems['hyp_xent'] = [[] for _ in range(batch_size)]

# at some time step
for batch_idx in range(len(hyp_xent)):
  if cant_skip[batch_idx]:
    mems['hyp_xent'][batch_idx].append(hyp_xent[batch_idx])
```

Comment (mrdrozdov) [reply python/spinn/fat_stack.py]: Then later we get entire array of `hyp_xent` for example i in some batch like this:

```
hyp_xent = mems['hyp_xent'][batch_idx]
```

Can fill with `None` for timesteps that are skipped, or might not even be a problem.

File: python/spinn/fat_stack.py [hunk 4]
```diff
@@ -570,6 +608,10 @@ def __call__(self, sentences, transitions, y_batch=None, train=True,
         accum_loss = self.classifier(y, Variable(y_batch, volatile=not train), train)
         self.accuracy = self.accFun(y, self.__mod.array(y_batch))
 
+        if self.use_reinforce:
+            rewards = - np.array([float(F.softmax_cross_entropy(y[i:(i+1)], y_batch[i:(i+1)]).data) for i in range(y_batch.shape[0])])
```
Comment (mrdrozdov) [inline python/spinn/fat_stack.py:610]: Making sure I understand. This gets the reward per example, rather than the mean?

Comment (mrdrozdov) [reply python/spinn/fat_stack.py:610]: You slice like `y[i:(i+1)]` to retain the dimensions? Nitpick, but can also do something like:

```
ys = F.split_axis(y, y.shape[0], axis=0, ...)
y_batches = F.split_axis(y_batch, y_batch.shape[0], axis=0, ...)
np.array([float(F.softmax_cross_entropy(yy, yy_batch) for yy, yy_batch in zip(ys, y_batches)])
```

File: python/spinn/fat_stack.py [hunk 5]
```diff
@@ -277,9 +286,11 @@ def run(self, print_transitions=False, run_internal_parser=False,
 
                             cant_skip_mask = np.tile(np.expand_dims(cant_skip, axis=1), (1, 2))
                             hyp_xent = F.split_axis(transition_hyp, transition_hyp.shape[0], axis=0)
-                            hyp_xent = F.concat([hyp_xent[i] for i, y in enumerate(cant_skip) if y], axis=0)
+                            hyp_xent = F.concat([hyp_xent[iii] for iii, y in enumerate(cant_skip) if y], axis=0)
```
Comment (mrdrozdov) [inline python/spinn/fat_stack.py:288]: Yikes. Good catch.

File: python/spinn/fat_stack.py [hunk 6]
```diff
@@ -373,6 +384,35 @@ def run(self, print_transitions=False, run_internal_parser=False,
...
+        statistics = [
+            F.squeeze(F.concat([F.expand_dims(ss, 1) for ss in s], axis=0))
+            if isinstance(s[0], Variable) else
+            np.array(reduce(lambda x, y: x + y.tolist(), s, []))
+            for s in statistics]
+
+        hyp_acc, truth_acc, hyp_xent, truth_xent = statistics
+
+        # Expand rewards
+        rewards = expandAlong(rewards, self.transition_mask)
```
Comment (mrdrozdov) [inline python/spinn/fat_stack.py:399]: Again, for my understanding, this sets rewards for skips to 0?

Comment (alexandresablayrolles) [reply python/spinn/fat_stack.py:399]: Yes. rewards is a (batch_size,) array, and transition_mask a (batch_size, seq_len) array, so expandAlong() populates a (batch_size, seq_len) with the rewards, and extracts only the ones when transition_mask is 1 (so rewards for skips are not in the result of expandAlong)
